### PR TITLE
Direct Messages (Private Chat)

### PR DESCRIPTION
**Pull request by Thomas-Frew**
Servers and clients form two-way connections and can propagate messages between each other, either in a directed or a flooding way.
Currently supported messages:

- HELLO
- PUBLIC_CHAT (public_chat "message" in the CLI)
- CLIENT_LIST
- CLIENT_LIST_REQUEST (online_list in the CLI)
- CLIENT_UPDATE
- CLIENT_UPDATE_REQUEST
- PRIVATE_CHAT (chat in the CLI)

Current (flawed) assumptions:

- All servers use the same certificate. (neighbourhood.olaf needs an upgrade).
- All messages are valid (signatures are not validated).

### CODE REVIEW

**Review by mortonjack — APPROVED** [review]
Looks good! Happy to merge as is. Left some comments you can address first if you please.

File: src/cli.hpp [hunk 1]
```diff
@@ -43,6 +43,42 @@ inline void cli(Connection &&connection, Client &&client,
...
+
+            std::getline(text_stream, text);
+
+            auto message_data = std::make_unique<PrivateChatData>(
+                std::move(servers), "0", std::move(symm_keys),
+                std::move(participants), text);
+
+            Message message{MessageType::PRIVATE_CHAT, std::move(message_data)};
+
+            nlohmann::json message_json = message.to_json();
```
[the remark this answers is not in the document]
Comment (Thomas-Frew) [reply src/cli.hpp:79]: Reasonable

File: src/messagehandler.cpp [hunk 2]
```diff
@@ -41,6 +44,17 @@ auto MessageHandler::handle_public_chat(Message &&message) const -> void {
...
 }
 
+auto MessageHandler::handle_private_chat(Message &&message) const -> void {
+    if (!this->verify_message(message)) {
+        return;
+    }
+
+    const auto &data = static_cast<const PrivateChatData &>(message.data());
+
+    std::cout << "[PRIVATE_CHAT] " << data.participants().at(0) << ": "
```
[the remark this answers is not in the document]
Comment (Thomas-Frew) [reply src/messagehandler.cpp:54]: If you are decrypting the message with a public key you can make sure the public key you just decrypted with matches the participant in the 0th position right?

File: src/messages.hpp [hunk 3]
```diff
@@ -89,8 +91,8 @@ class ClientListData : public MessageData {
         std::map<std::string_view, std::vector<std::string_view>> &&online_list)
         : m_online_list(std::move(online_list)) {}
 
-    static auto
-    from_json(const nlohmann::json &j) -> std::unique_ptr<ClientListData>;
+    static auto from_json(const nlohmann::json &j)
+        -> std::unique_ptr<ClientListData>;
```
[the remark this answers is not in the document]
Comment (Thomas-Frew) [reply src/messages.hpp:95]: All's well that clangs well

File: server/server.py [hunk 4]
```diff
@@ -327,7 +324,38 @@ async def handle_client_update(self, message_data):
...
+        destination_clients = chat_data.get('participants')
+        destination_servers = message_data.get('destination_servers')
+        unique_destinations = list(set(destination_servers))
+        
+        # The first participant (sender) always gets their message reflected back to them
+        sender_pub_key = destination_clients[0]
+        if self.clients.get(sender_pub_key) is not None:
+            await self.clients[sender_pub_key].send(json.dumps(message))
+        
+        # Propgate the message locally to all recieving users
```
[the remark this answers is not in the document]
Comment (Thomas-Frew) [reply server/server.py:342]: I will fix this later...

**Pull request merged**